src/object/obj_tx.c

```diff
@@ -47,12 +47,23 @@
 #define DTX_SUB_REQ_MAX		((1ULL << 32) - 1)
 #define DTX_SUB_REQ_DEF		16
 
+/**
+ * Client transaction status
+ *
+ * If a TX operation encounters
+ *
+ *   - -DER_EXIST or -DER_NONEXIST, the TX remains OPEN;
+ *   - a retryable operation error, the TX remains OPEN;
+ *   - -DER_TX_RESTART or -DER_STALE, the TX becomes FAILED; or
+ *   - an error not listed above, the TX becomes FAILED or ABORTED (currently
```

Comment:
Why do we want to treat failures differently?
DER_EXIST/NOEXIST vs something else like DER_NO_PERM for example?
I guess i don't get why the latter failures should mark the entire TX as failed

Reply:
-DER_EXIST or _DER_NONEXIST are results of read operations, not real run-time errors indicating something is wrong. E.g., callers may do "if (rc == -DER_NONEXIST) {do_something_else and tx_commit}".

For a returnable operation error, the operation is effectively still in progress.

For another error like -DER_NO_PERM, since we don't have ACLs for objects/keys/values, it must be a run-time error indicating something is wrong. So at least the current generation of the TX should stop. Whether we should allow restarting a new generation of this TX is up to debate depending on the specific error code, I think.

Reply:
On second thought, you're right. For the last class "an error not listed above", maybe we could leave the TX OPEN...